File: web/src/Entity/Produit.php
<?php

namespace App\Entity;

use App\Repository\ProduitRepository;
use Doctrine\ORM\Mapping as ORM;
use Gedmo\Mapping\Annotation as Gedmo;

#[ORM\Entity(repositoryClass: ProduitRepository::class)]
#[ORM\Table(name: 'produit')]
class Produit
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(name: 'nom_produit', length: 255)]
    private ?string $nomProduit = null;

    #[ORM\Column(length: 255)]
    private ?string $adresse = null;

    #[ORM\Column(type: 'float')]
    private ?float $prix = null;

    #[ORM\Column]
    private ?int $quantite = null;

    #[ORM\Column(name: 'image_url', length: 500, nullable: true)]
    private ?string $imageURL = null;

    #[Gedmo\Slug(fields: ['nomProduit'])]
    #[ORM\Column(length: 255, unique: true)]
    private ?string $slug = null;

    #[ORM\ManyToOne(inversedBy: 'produits')]
    #[ORM\JoinColumn(name: 'categorie_id', referencedColumnName: 'id', nullable: false)]
    private ?Categorie $categorie = null;

    #[ORM\ManyToOne(inversedBy: 'produits')]
    #[ORM\JoinColumn(name: 'fournisseur_id', referencedColumnName: 'id', nullable: true, onDelete: 'SET NULL')]
    private ?User $fournisseur = null;

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getNomProduit(): ?string
    {
        return $this->nomProduit;
    }

    public function setNomProduit(string $nomProduit): static
    {
        $this->nomProduit = $nomProduit;
        return $this;
    }

    public function getAdresse(): ?string
    {
        return $this->adresse;
    }

    public function setAdresse(string $adresse): static
    {
        $this->adresse = $adresse;
        return $this;
    }

    public function getPrix(): ?float
    {
        return $this->prix;
    }

    public function setPrix(float $prix): static
    {
        $this->prix = $prix;
        return $this;
    }

    public function getQuantite(): ?int
    {
        return $this->quantite;
    }

    public function setQuantite(int $quantite): static
    {
        $this->quantite = $quantite;
        return $this;
    }

    public function hasAvailableQuantity(int $quantite): bool
    {
        return $quantite > 0 && ($this->quantite ?? 0) >= $quantite;
    }

    public function decreaseQuantite(int $quantite): static
    {
        $this->quantite = max(0, ($this->quantite ?? 0) - $quantite);
        return $this;
    }

    public function getImageURL(): ?string
    {
        return $this->imageURL;
    }

    public function setImageURL(?string $imageURL): static
    {
        $this->imageURL = $imageURL;
        return $this;
    }

    public function getSlug(): ?string
    {
        return $this->slug;
    }

    public function setSlug(?string $slug): static
    {
        $this->slug = $slug;
        return $this;
    }

    public function getCategorie(): ?Categorie
    {
        return $this->categorie;
    }

    public function setCategorie(?Categorie $categorie): static
    {
        $this->categorie = $categorie;
        return $this;
    }

    public function getFournisseur(): ?User
    {
        return $this->fournisseur;
    }

    public function setFournisseur(?User $fournisseur): static
    {
        $this->fournisseur = $fournisseur;
        return $this;
    }
}

File: JavaFX/src/main/java/models/Produit.java
package models;

public class Produit {

    private int id;
    private String nomProduit;
    private String adresse;
    private double prix;
    private int quantite;
    private String imageURL;
    private String slug;
    private int categorieId;
    private String categorieNom;
    private Integer fournisseurId;
    private String fournisseurNom;

    public Produit(int id, String nomProduit, String adresse, double prix, int quantite, String imageURL) {
        this(id, nomProduit, adresse, prix, quantite, imageURL, null, 0, null, null, null);
    }

    public Produit(int id, String nomProduit, String adresse, double prix, int quantite, String imageURL,
                   String slug, int categorieId, String categorieNom, Integer fournisseurId, String fournisseurNom) {
        this.id = id;
        this.nomProduit = nomProduit;
        this.adresse = adresse;
        this.prix = prix;
        this.quantite = quantite;
        this.imageURL = imageURL;
        this.slug = slug;
        this.categorieId = categorieId;
        this.categorieNom = categorieNom;
        this.fournisseurId = fournisseurId;
        this.fournisseurNom = fournisseurNom;
    }

    public Produit(String nomProduit, String adresse, double prix, int quantite, String imageURL) {
        this.nomProduit = nomProduit;
        this.adresse = adresse;
        this.prix = prix;
        this.quantite = quantite;
        this.imageURL = imageURL;
    }

    public int getId() { return id; }
    public String getNomProduit() { return nomProduit; }
    public String getAdresse() { return adresse; }
    public double getPrix() { return prix; }
    public int getQuantite() { return quantite; }
    public String getImageURL() { return imageURL; }
    public String getSlug() { return slug; }
    public int getCategorieId() { return categorieId; }
    public String getCategorieNom() { return categorieNom; }
    public Integer getFournisseurId() { return fournisseurId; }
    public String getFournisseurNom() { return fournisseurNom; }

    public void setId(int id) { this.id = id; }
    public void setNomProduit(String nomProduit) { this.nomProduit = nomProduit; }
    public void setAdresse(String adresse) { this.adresse = adresse; }
    public void setPrix(double prix) { this.prix = prix; }
    public void setQuantite(int quantite) { this.quantite = quantite; }
    public void setImageURL(String imageURL) { this.imageURL = imageURL; }
    public void setSlug(String slug) { this.slug = slug; }
    public void setCategorieId(int categorieId) { this.categorieId = categorieId; }
    public void setCategorieNom(String categorieNom) { this.categorieNom = categorieNom; }
    public void setFournisseurId(Integer fournisseurId) { this.fournisseurId = fournisseurId; }
    public void setFournisseurNom(String fournisseurNom) { this.fournisseurNom = fournisseurNom; }

    @Override
    public String toString() {
        return "Produit{" +
                "id=" + id +
                ", nomProduit='" + nomProduit + '\'' +
                ", adresse='" + adresse + '\'' +
                ", prix=" + prix +
                ", quantite=" + quantite +
                ", imageURL='" + imageURL + '\'' +
                ", categorieId=" + categorieId +
                ", categorieNom='" + categorieNom + '\'' +
                ", fournisseurId=" + fournisseurId +
                '}';
    }
}
